File: ObserverPattern/Observer.cs
using System;
using System.Collections.Generic;

public class BlogPost {
    public string Data { get; set; }
    public BlogPost(string data) => Data = data;
}

public class WeeklyNewsletter {
    public string Data { get; set; }
    public WeeklyNewsletter(string data) => Data = data;
}

public enum NotifyEvents {
    BlogPost,
    WeeklyNewsletter
}

public interface ISubscriber {
    void Update(NotifyEvents notifyEvent, string msg);
}

public class Subscriber : ISubscriber {
    public string Name { get; set; }
    public Subscriber(string name) => Name = name;

    public void Update(NotifyEvents notifyEvent, string msg) {
        Console.WriteLine($"Notifying {Name} about a new {notifyEvent}: {msg}");
    }
}

public class BlogManagement {
    private List<BlogPost> _blogPosts = new();
    private List<WeeklyNewsletter> _weeklyNewsletters = new();
    private Dictionary<NotifyEvents, List<ISubscriber>> _eventSubscribers;

    public BlogManagement() {
        _eventSubscribers = new Dictionary<NotifyEvents, List<ISubscriber>> {
            { NotifyEvents.BlogPost,         new List<ISubscriber>() },
            { NotifyEvents.WeeklyNewsletter, new List<ISubscriber>() }
        };
    }

    public void Subscribe(NotifyEvents notifyEvent, ISubscriber person) {
        _eventSubscribers[notifyEvent].Add(person);
    }

    public void Unsubscribe(NotifyEvents notifyEvent, ISubscriber person) {
        _eventSubscribers[notifyEvent].Remove(person);
    }

    public void AddPost(BlogPost post) {
        _blogPosts.Add(post);
        Notify(NotifyEvents.BlogPost, post.Data);
    }

    public void AddWeeklyNewsletter(WeeklyNewsletter newsletter) {
        _weeklyNewsletters.Add(newsletter);
        Notify(NotifyEvents.WeeklyNewsletter, newsletter.Data);
    }

    private void Notify(NotifyEvents notifyEvent, string msg) {
        foreach (var person in _eventSubscribers[notifyEvent]) {
            person.Update(notifyEvent, msg);
        }
    }
}

class Program {
    static void Main() {
        var management = new BlogManagement();

        var ahmed = new Subscriber("Ahmed");
        var sara  = new Subscriber("Sara");

        management.Subscribe(NotifyEvents.BlogPost, ahmed);
        management.Subscribe(NotifyEvents.BlogPost, sara);
        management.Subscribe(NotifyEvents.WeeklyNewsletter, ahmed);

        management.AddPost(new BlogPost("Design Patterns 101"));
        management.AddWeeklyNewsletter(new WeeklyNewsletter("Weekly digest #1"));

        management.Unsubscribe(NotifyEvents.BlogPost, sara);
        management.AddPost(new BlogPost("Observer Pattern deep dive"));
    }
}

File: ObserverPattern/observer.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <string>
using namespace std;

class BlogPost {
public:
    string data;
    BlogPost(string data) : data(data) {}
};

class WeeklyNewsletter {
public:
    string data;
    WeeklyNewsletter(string data) : data(data) {}
};

enum NotifyEvents {
    BlogPostEvent,
    WeeklyNewsletterEvent
};

class ISubscriber {
public:
    virtual ~ISubscriber() = default;
    virtual void update(NotifyEvents event, string msg) = 0;
};

class Subscriber : public ISubscriber {
public:
    string name;
    Subscriber(string name) : name(name) {}

    void update(NotifyEvents event, string msg) override {
        cout << "Notifying " << name << " about: " << msg << endl;
    }
};

class BlogManagement {
    vector<BlogPost> blogPosts;
    vector<WeeklyNewsletter> weeklyNewsLetters;
    map<NotifyEvents, vector<ISubscriber*>> eventSubscribers;

public:

    void subscribe(NotifyEvents event, ISubscriber* person) {
        eventSubscribers[event].push_back(person);
    }

    void unsubscribe(NotifyEvents event, ISubscriber* person) {
        auto& list = eventSubscribers[event];
        list.erase(remove(list.begin(), list.end(), person), list.end());
    }

    void addPost(BlogPost newPost) {
        blogPosts.push_back(newPost);
        notify(BlogPostEvent, "New blog post: " + newPost.data);
    }

    void addWeeklyNewsletter(WeeklyNewsletter newsletter) {
        weeklyNewsLetters.push_back(newsletter);
        notify(WeeklyNewsletterEvent, "New newsletter: " + newsletter.data);
    }

private:
    void notify(NotifyEvents event, string msg) {
        for (auto person : eventSubscribers[event]) {
            person->update(event, msg);
        }
    }
};

int main() {
    BlogManagement management;

    Subscriber ahmed("Ahmed");
    Subscriber sara("Sara");

    management.subscribe(BlogPostEvent, &ahmed);
    management.subscribe(BlogPostEvent, &sara);
    management.subscribe(WeeklyNewsletterEvent, &ahmed);

    management.addPost(BlogPost("Design Patterns 101"));
    management.addWeeklyNewsletter(WeeklyNewsletter("Weekly digest #1"));

    management.unsubscribe(BlogPostEvent, &sara);
    management.addPost(BlogPost("Observer Pattern deep dive"));

    return 0;
}
